cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/KafkaAdminTopicConfigProvider.java

```diff
@@ -0,0 +1,168 @@
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import com.linkedin.kafka.cruisecontrol.config.constants.ExecutorConfig;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.Config;
+import org.apache.kafka.clients.admin.ConfigEntry;
+import org.apache.kafka.common.config.ConfigResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The Kafka topic config provider implementation based on files. The format of the file is JSON, listing properties:
+ * <pre>
+ *   {
+ *     "min.insync.replicas": 1,
+ *     "an.example.cluster.config": false
+ *   }
+ * </pre>
+ *
+ */
+public class KafkaAdminTopicConfigProvider implements TopicConfigProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaAdminTopicConfigProvider.class);
+
+  public static final String CLUSTER_CONFIGS_FILE = "cluster.configs.file";
+  private static Properties _clusterConfigs;
+  private AdminClient _adminClient;
+  private long _adminTimeoutMs;
+
+  @Override
+  //TODO: This is just returning a parsed file? Why not the cluster config from the actual cluster?
+  public Properties clusterConfigs() {
+    return _clusterConfigs;
+  }
+
+  @Override
+  public Properties topicConfigs(String topic) {
+    Config topicConfig = null;
+    ConfigResource topicResource = new ConfigResource(ConfigResource.Type.TOPIC, topic);
+    try {
+      LOG.debug("Requesting details for topic '{}'", topic);
+      topicConfig = _adminClient
+              .describeConfigs(Collections.singletonList(topicResource))
+              .all()
+              .get(_adminTimeoutMs, TimeUnit.MILLISECONDS)
+              .get(topicResource);
+    } catch (InterruptedException e) {
+      LOG.error("The request for the configuration of topic '{}' was interrupted", topic);
+      e.printStackTrace();
+    } catch (ExecutionException e) {
+      LOG.error("The request for the configuration of topic '{}' failed", topic);
+      e.printStackTrace();
+    } catch (TimeoutException e) {
+      LOG.error("The request for the configuration of topic '{}' timed out", topic);
+      e.printStackTrace();
+    }
+
+    if (topicConfig != null) {
+      return convertTopicConfigToProperties(topicConfig);
+    } else {
+      LOG.error("The configuration for topic '{}' could not be retrieved", topic);
+      return new Properties();
+    }
+  }
+
+  @Override
+  public Map<String, Properties> allTopicConfigs() {
+    Map<ConfigResource, Config> topicConfigs = null;
+    try {
+      LOG.debug("Requesting configurations for all topics");
+      topicConfigs = _adminClient
+              .listTopics()
+              .names()
+              .thenApply(
+                      topicNameSet -> _adminClient.describeConfigs(
+                              topicNameSet.stream().map(name -> new ConfigResource(ConfigResource.Type.TOPIC, name)).collect(Collectors.toList())
+                      ).all()
```

Comment:
`all()` method returns a future which succeeds _only if all the config descriptions succeed_.
To prevent individual failures causing the whole request to fail, can we instead use `values()` response and handle individual failures separately?

Reply:
Good point, I think I have changed this to something that should work but it might be a bit _over_ engineered. Let me know what you think.